src/test/txindex_tests.cpp

```diff
@@ -0,0 +1,47 @@
+// Copyright (c) 2017 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <index/txindex.h>
+#include <script/standard.h>
+#include <test/test_bitcoin.h>
+#include <util.h>
+#include <utiltime.h>
+#include <validation.h>
+
+#include <boost/test/unit_test.hpp>
+
+BOOST_AUTO_TEST_SUITE(txindex_tests)
+
+BOOST_FIXTURE_TEST_CASE(txindex_initial_sync, TestChain100Setup)
+{
+    TxIndex txindex(MakeUnique<TxIndexDB>(1 << 20, true));
+    txindex.Start();
+
+    // Allow tx index to catch up with the block index.
+    constexpr int64_t timeout_ms = 10 * 1000;
+    int64_t time_start = GetTimeMillis();
+    while (!txindex.BlockUntilSyncedToCurrentChain()) {
+        BOOST_REQUIRE(time_start + timeout_ms > GetTimeMillis());
+        MilliSleep(100);
+    }
+
+    for (const auto& txn : coinbaseTxns) {
+        CDiskTxPos postx;
+        BOOST_CHECK(txindex.FindTx(txn.GetHash(), postx));
+    }
+
+    for (int i = 0; i < 10; i++) {
+        CScript coinbase_script_pub_key = GetScriptForDestination(coinbaseKey.GetPubKey().GetID());
+        std::vector<CMutableTransaction> no_txns;
+        const CBlock& block = CreateAndProcessBlock(no_txns, coinbase_script_pub_key);
+        const CTransaction& txn = *block.vtx[0];
+
+        txindex.BlockUntilSyncedToCurrentChain();
+
+        CDiskTxPos actual_postx;
+        BOOST_CHECK(txindex.FindTx(txn.GetHash(), actual_postx));
```

Comment:
I added checks for 1 & 3. With regards to persistence, the DB in the test environment currently uses the in-memory LevelDB environment rather than actually writing to disk, which makes cleanup nicer and probably makes the test marginally faster.

How important do you think it is to test the persistence at this layer? My feeling is that it's probably not adding much since it's just using the common CDBWrapper abstraction.